gossip/emitter/throttling/throttling.go

```diff
@@ -0,0 +1,132 @@
+
+// ThrottlingState manages the state for event emission throttling based on
+// the set of dominant validators.
+// This type contains the state needed to decide whether to skip event emission
+// for a given validator, based on its stake and the stake distribution among
+// all validators, and properties of the events to be emitted.
+type ThrottlingState struct {
+	// throttler configuration parameters
+	thisValidatorID                     idx.ValidatorID
+	dominatorsThreshold                 float64
+	maxSkippedEventsWithSameFrameNumber uint
+
+	// means to access the world state
+	world WorldReader
+
+	// accumulated state
+	lastUsedFrame             idx.Frame
+	lastEmissionBlockNumber   idx.Block
+	currentSkippedEventsCount uint
+}
+
+type WorldReader interface {
+	GetRules() opera.Rules
+	GetLatestBlockIndex() idx.Block
+	GetEpochValidators() (*pos.Validators, idx.Epoch)
+}
+
+// NewThrottlingState creates a new ThrottlingState for a given validator ID
+// and dominating percentile threshold.
+//
+// dominatingPercentile parameter specifies the fraction of total stake
+// that defines the dominant set of validators. For example, a value of 0.75
+// means that the dominant set is the smallest set of validators whose combined
+// stake is at least 75% of the total stake.
+//
+// stalledFrameTimeout parameter specifies the duration to consider an event
+// as stalled in the current frame.
+func NewThrottlingState(
+	validatorID idx.ValidatorID,
+	dominatingPercentile float64,
+	maxSkippedEventsWithSameFrameNumber uint,
+	stateReader WorldReader,
+) *ThrottlingState {
+	return &ThrottlingState{
+		thisValidatorID:                     validatorID,
+		dominatorsThreshold:                 min(max(dominatingPercentile, 0.7), 1),
+		maxSkippedEventsWithSameFrameNumber: maxSkippedEventsWithSameFrameNumber,
+		world:                               stateReader,
+	}
+}
+
+// SkipEventEmission determines whether to skip the emission of the given event.
+//
+// It returns true if the event emission should be skipped, false otherwise.
+func (ts *ThrottlingState) SkipEventEmission(event inter.EventPayloadI) bool {
+	skip := ts.skipEvent(event)
+	ts.lastUsedFrame = event.Frame()
+
+	if skip {
+		ts.currentSkippedEventsCount++
+	} else {
+		ts.lastEmissionBlockNumber = ts.world.GetLatestBlockIndex()
+		ts.currentSkippedEventsCount = 0
+	}
+	return skip
+}
+
+func (ts *ThrottlingState) skipEvent(event inter.EventPayloadI) bool {
+	// Do not skip emission if the event carries transactions
+	if len(event.Transactions()) > 0 {
+		return false
+	}
+
+	rules := ts.world.GetRules()
+
+	// Do not skip emission if the event is in the same frame as the last emitted event
+	// for a given period of time.
+	// This means that there no progress in the network can be observed. The stake of this
+	// validator stake may be needed to reach quorum in the current frame.
+	if ts.lastUsedFrame == event.Frame() &&
```

Comment:
could the frame have advanced without `currentSkippedEventsCounts` being reseted ?